servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ReplayStrategies.java

```diff
@@ -0,0 +1,244 @@
+         * are visible from the consumer thread and we never "miss" a timer schedule event if the queue becomes empty.
+         */
+        private volatile long stateSize;
+        @Nullable
+        private volatile Cancellable timerCancellable;
+
+        MostRecentTimeLimitedReplayAccumulator(final int maxItems, final Duration ttl, final Executor executor) {
+            if (ttl.isNegative()) {
+                throw new IllegalArgumentException("ttl: " + ttl + "(expected non-negative)");
+            }
+            if (maxItems <= 0) {
+                throw new IllegalArgumentException("maxItems: " + maxItems + "(expected >0)");
+            }
+            this.executor = requireNonNull(executor);
+            this.ttlNanos = ttl.toNanos();
+            this.maxItems = maxItems;
+            items = new ConcurrentLinkedQueue<>(); // SpMc
+        }
+
+        @Override
+        public void accumulate(@Nullable final T t) {
+            // We may exceed max items in the queue but this method isn't invoked concurrently, so we only go over by
+            // at most 1 item.
+            items.add(new TimeStampSignal<>(executor.currentTime(NANOSECONDS), t));
+            for (;;) {
+                final long currentStateSize = stateSize;
+                final int currentSize = getSize(currentStateSize);
+                final int nextState = getState(currentStateSize) + 1;
+                if (currentSize >= maxItems) {
+                    if (stateSizeUpdater.compareAndSet(this, currentStateSize,
+                            buildStateSize(nextState, currentSize))) {
+                        items.poll();
+                        break;
+                    }
+                } else if (stateSizeUpdater.compareAndSet(this, currentStateSize,
+                        buildStateSize(nextState, currentSize + 1))) {
+                    if (currentSize == 0) {
+                        schedulerTimer(ttlNanos);
+                    }
+                    break;
+                }
+            }
+        }
+
+        @Override
+        public void deliverAccumulation(final Consumer<T> consumer) {
+            for (TimeStampSignal<T> timeStampSignal : items) {
+                consumer.accept(timeStampSignal.signal);
+            }
+        }
+
+        @Override
+        public void cancelAccumulation() {
+            final Cancellable cancellable = timerCancellableUpdater.getAndSet(this, EMPTY_SUBSCRIPTION_NO_THROW);
+            if (cancellable != null) {
+                cancellable.cancel();
+            }
+        }
+
+        private static int getSize(long stateSize) {
+            return (int) stateSize;
+        }
+
+        private static int getState(long stateSize) {
+            return (int) (stateSize >>> 32);
+        }
+
+        private static long buildStateSize(int state, int size) {
+            return (((long) state) << 32) | size;
+        }
+
+        private void schedulerTimer(long nanos) {
```

Comment:
In other similar cases (e.g. `buffer(..)`) we also require a timer. If you just limit based upon time you may fall into the trap of too many items during your time interval, so to remove potential sharp edges by default we require both. You can simulate a "time only" option by setting the limit very high.

Reply:
Sorry, I wasn't terribly clear, I think we're on the same page for `maxItems` but not the value of the timer _task_. I know we have two different limiters and I think the `maxItems` is the overwhelmingly most valuable of the two limits and a must have.
The timer task, otoh, doesn't appear to buy a lot other than (theoretically) better GC behavior due to the task eagerly evicting expired items: we still have to potentially buffer `maxItems` that come in rapidly so the max memory usage is theoretically the same. However, even without the timer task we can still honor the time constraint by checking if items are expired when it's time to emit them and pruning then. By pruning during emission we save us the overhead of the timer tasks themselves and some extra concurrency primitive overhead since the emission happens inside the protection of our existing concurrency primitives.

Reply:
Ah I see what you are saying. I'm not sure there is a "right answer" and depends upon the scenario.

Timer Task Desirable:
Desired lifetime of element is `<small number>` (e.g. 1 second), an upper bound of `<high number>`, data is expected to arrive on average in 100ms intervals (but can be bursty). I don't want to retain `high number` in memory at all times and the timer is expected to dictate queue size in most cases.

Timer Task is Extra Overhead:
Desired lifetime of element is `<high number>` (e.g. 15 seconds), an upper bound of `<small number>`, data is expected to arrive on average in 100ms intervals (but may have gaps in time). The timer isn't really useful here (but depending upon the granularity may not also be a primary resource consumer).

We could always provide an alternative strategy/option if we want to but rational is the timer task based approach provides a more conservatively bounded approach to start (although more complex from implementation perspective).

Reply:
I added a `historyTtlBuilder` overload that has a `lazy` argument to demonstrate the two approaches. The method is currently package private and I'm not sure we want to expose both yet